core/src/main/java/org/infinispan/container/offheap/BoundedOffHeapDataContainer.java

```diff
@@ -259,8 +259,8 @@ private void ensureSize() {
             try {
                InternalCacheEntry<WrappedBytes, WrappedBytes> ice = offHeapEntryFactory.fromMemory(addressToRemove);
                passivator.passivate(ice);
-               // TODO: this reareads the object from memory again!
-               performRemove(addressToRemove, ice.getKey());
+               // TODO: this rereads the object from memory again!
+               performRemove(memoryLookup.getMemoryAddress(ice.getKey()), ice.getKey());
```

Comment:
This is the actual big bug fix. The problem was this was passing the address pointer to remove (which would be a fine assumption from the name as Dan mentioned). However this method required the address of the first node in the memory lookup. This is required since this is a forward linked list only and such when we remove an element we need its parent to mend the linked list. By not passing this the list would be corrupted which could cause all sorts of issues (especially if this memory location was reused by another entry - which was almost always the case)